File: bot.js
var Libhoney = require("libhoney");
var hny = new Libhoney({
    writeKey: process.env.HONEYCOMB_KEY,
    dataset: "crabravebot"
})
const { Composer } = require('micro-bot');
const child_process = require('child_process');
const fs = require('fs');
const url = require('url');
const { JSDOM } = require("jsdom");

// the bot
const bot = new Composer();

// start command
bot.command('/start', async ({ from, replyWithMarkdown, botInfo }) =>
  replyWithMarkdown(`Hi *${from.first_name || from.username}*!
To shitpost, type @${botInfo.username} and type the text you want to overlay over crab rave.
This was made by @boringcactus in one afternoon when she was bored.
This bot isn't super reliable but the source is at https://glitch.com/edit/#!/${process.env.PROJECT_DOMAIN}`));

// styles
const STYLES = {
  'classic': {
    'video': 'https://cdn.glitch.com/70ea37b5-d264-46e5-a1db-29c786c86515%2FCrabRaveQuieter.mp4?v=1548223581701',
    'thumb': 'https://cdn.glitch.com/70ea37b5-d264-46e5-a1db-29c786c86515%2FCrabRaveBackground.png?v=1566350372048',
  },
  'otamatone': {
    'video': 'https://cdn.glitch.com/70ea37b5-d264-46e5-a1db-29c786c86515%2FCrabRaveOtamatone.mp4?v=1566354239409',
    'thumb': 'https://cdn.glitch.com/70ea37b5-d264-46e5-a1db-29c786c86515%2FCrabRaveOtamatoneBackground.png?v=1566354702285',
  },
}

// inline query
bot.on('inline_query', async ({ inlineQuery, answerInlineQuery }) => {
  const query = inlineQuery.query || '';
  console.log('Got query', query);
  if (query.length > 1) {
    const result = [
      {
        type: "video",
        id: "classic",
        video_url: "https://" + process.env.PROJECT_DOMAIN + ".glitch.me/video/" + encodeURIComponent(query) + ".mp4?v1",
        mime_type: "video/mp4",
        thumb_url: "https://" + process.env.PROJECT_DOMAIN + ".glitch.me/video/" + encodeURIComponent(query) + ".png?v1",
        title: 'Classic'
      },
      {
        type: "video",
        id: "otamatone",
        video_url: "https://" + process.env.PROJECT_DOMAIN + ".glitch.me/video/" + encodeURIComponent(query) + ".mp4?v1&style=otamatone",
        mime_type: "video/mp4",
        thumb_url: "https://" + process.env.PROJECT_DOMAIN + ".glitch.me/video/" + encodeURIComponent(query) + ".png?v1&style=otamatone",
        title: 'Otamatone (original by TheRealSullyG)'
      },
    ];
    return answerInlineQuery(result);
  }
});

function fixPaths() {
  for (let style of Object.keys(STYLES)) {
    try {
      let path = require('path').join('/tmp', style, 'video');
      fs.mkdirSync(path, {recursive: true});
    } catch (e) {
      console.log(e);
    }
  }
}

module.exports = {
  bot,
  server(req, res) {
    if (req.url === '/') {
      res.end(`
        <html>
        <head>
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <title>Crab Rave Bot</title>
        </head>
        <body>
        <main>
        <h1>it's a tool for adding things to crab rave</h1>
        <form action="/add-text" method="GET">
        <textarea id="text" name="text" rows="6" cols="25"></textarea>
        <p>
          Style:
          <input type="radio" id="classic" name="style" value="classic" checked><label for="classic">Classic</label>
          <input type="radio" id="otamatone" name="style" value="otamatone"><label for="otamatone">Otamatone</label> (<a href="https://youtu.be/VI5I3MuKJlo">original</a> by TheRealSullyG)
        </p>
        <input type="submit" value="Overlay!">
        </form>
        <a href="https://t.me/crabravebot">also available as a Telegram bot</a>
        </main>
        <img id="preview" style="max-width: 100%;" src="https://${process.env.PROJECT_DOMAIN}.glitch.me/video/.png">
        <script type="text/javascript">
        const img = document.getElementById('preview'),
          text = document.getElementById('text'),
          classic = document.getElementById('classic');
        setInterval(() => {
          let style = classic.checked ? 'classic' : 'otamatone';
          img.src = "https://${process.env.PROJECT_DOMAIN}.glitch.me/video/" + encodeURIComponent(text.value) + '.png?style=' + style;
        }, 1000);
        </script>
        </body>
        </html>
      `);
    } else if (req.url.startsWith('/add-text')) {
      const parsed = url.parse(req.url, true);
      const data = parsed.query.text;
      const style = parsed.query.style;
      res.writeHead(303, {
        'Location': "https://" + process.env.PROJECT_DOMAIN + ".glitch.me/video/" + encodeURIComponent(data) + ".mp4?style=" + style
      });
      res.end();
    } else if (req.url.startsWith('/video/')) {
      fixPaths();
      if (!fs.existsSync('/tmp/classic/video')) {
        console.log('what???');
        child_process.spawnSync('ls', ['/tmp', '/tmp/classic', '/tmp/classic/video']);
      }
      const parsed = url.parse(req.url, true);
      const match = /^\/video\/(.*)\.(mp4|png)/.exec(parsed.pathname);
      if (match === null) {
        console.error('Bad URL: ' + match);
        res.writeHead(400);
        res.end();
        return;
      }
      const text = decodeURIComponent(match[1]);
      console.log('Match:', text);
      const type = match[2];
      if (type === 'mp4') {
        const ev = hny.newEvent();
        ev.addField('type', 'bake');
        ev.addField('text', text);
        ev.send();
      }
      const style = parsed.query.style || 'classic';
      if (!Object.keys(STYLES).includes(style)) {
        console.error("Bad style");
        res.writeHead(400);
        res.end();
        return;
      }
      const path = require('path').join('/tmp', style, encodeURIComponent(match[0]));
      if (type === 'mp4') {
        res.setHeader('Content-Type', 'video/mp4');
      } else {
        res.setHeader('Content-Type', 'image/png');
      }
      console.log('Name:', path);
      // TODO don't do this dumb thing
      if (!fs.existsSync(path)) {
        if (!fs.existsSync(path + '.png')) {
          const dom = new JSDOM(`<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<svg version="1.1" viewBox="0 0 848 480">
</svg>`, {contentType: 'image/svg+xml'});
          const document = dom.window.document;
          const lines = text.split('\n');
          let y = 300 - (75 * lines.length) / 2;
          for (let line of lines) {
            const lineNode = document.createElement('text');
            lineNode.setAttribute('x', 424);
            lineNode.setAttribute('y', y);
            y += 75;
            lineNode.setAttribute('style', "text-anchor:middle;alignment-baseline:middle;font-family:'DejaVu Sans',sans-serif;font-weight:bold;font-size:48pt;fill:white;stroke:black;stroke-width:1px;");
            lineNode.textContent = line;
            document.querySelector('svg').append(lineNode);
          }
          const svg = '<?xml version="1.0" encoding="UTF-8" standalone="no"?>\n' + dom.serialize();
          fs.writeFileSync(path + '.svg', svg);
          child_process.spawnSync('convert', ['-background', 'none', path + '.svg', path + '.png'], {stdio: ['inherit', 'inherit', 'inherit']});
        }
        if (type === 'mp4') {
          let background = STYLES[style].video;
          let args = [
            '-hide_banner',
            '-i', background,
            '-i', path + '.png',
            '-filter_complex', 'overlay=x=0:y=0',
            '-c:v', 'libx264', '-preset', 'superfast', '-crf', '27', '-f', 'mp4', '-c:a', 'copy',
            '-y', path
          ];
          console.log('ffmpeg', ...args);
          let child = child_process.spawnSync(
            'ffmpeg',
            args,
            {
              stdio: ['inherit', 'inherit', 'inherit']
            }
          );
        } else {
          let background_url = STYLES[style].thumb;
          let background = '/tmp/' + style + '.png';
          if (!fs.existsSync(background)) {
            child_process.spawnSync('curl', ['-o', background, background_url]);
          }
          child_process.spawnSync('composite', [path + '.png', background, path]);
        }
      }
      if (fs.existsSync(path)) {
        fs.createReadStream(path).pipe(res);
      } else {
        console.log('what the fuck', path);
      }
    }
  },
};

File: bot.py
import logging
import os
from io import BytesIO
from dataclasses import dataclass
from pathlib import Path
from urllib.parse import quote_plus as url_quote
import subprocess
import tempfile
from queue import Queue
from threading import Thread
import json

import telegram
from telegram.ext import Dispatcher, MessageHandler, CommandHandler, InlineQueryHandler, Filters, CallbackContext
from dotenv import load_dotenv, find_dotenv
from PIL import Image, ImageDraw, ImageFont
import libhoney
from flask import Flask, render_template, request
app = Flask(__name__)

load_dotenv(find_dotenv())

libhoney.init(writekey=os.environ.get('HONEYCOMB_KEY'), dataset="crabravebot", debug=True)

font = ImageFont.truetype("assets/fonts/NotoSans-Regular.ttf", 48)

logging.basicConfig(format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
                    level=logging.INFO)
logger = logging.getLogger(__name__)


@dataclass
class Style:
    """Crab rave style"""
    id: str
    name: str
    source: str
    image: Path
    video: Path


def get_styles():
    templates = Path('assets/templates')
    result = []
    for folder in templates.iterdir():
        s_id = folder.name
        with list(folder.glob('*.json'))[0].open('r') as meta:
            metadata = json.load(meta)
            name = metadata['name']
            source = metadata['source']
        image = list(folder.glob('*.png'))[0]
        video = list(folder.glob('*.mp4'))[0]
        result.append(Style(s_id, name, source, image, video))
    return result


STYLES = get_styles()


def word_wrap(text: str, draw: ImageDraw.ImageDraw, width: int) -> str:
    text_width, text_height = draw.multiline_textsize(text, font=font)
    if text_width > width:
        lines = text.splitlines()
        wrapped_text = []
        for line in lines:
            line_width, _ = draw.textsize(line, font=font)
            if line_width > width:
                new_line = ''
                last_line = new_line
                for word in line.split():
                    if len(new_line) > 0:
                        new_line += ' '
                    new_line += word
                    line_width, _ = draw.textsize(new_line, font=font)
                    if line_width > width:
                        wrapped_text.append(last_line)
                        new_line = word
                    last_line = new_line
                if len(last_line) > 0:
                    wrapped_text.append(last_line)
            else:
                wrapped_text.append(line)
        return '\n'.join(wrapped_text)


def render_text(text: str, base: Image):
    if base.mode == 'RGB':
        white = (255, 255, 255)
        black = (0, 0, 0)
    elif base.mode == 'RGBA':
        white = (255, 255, 255, 255)
        black = (0, 0, 0, 255)
    else:
        raise ValueError('Base image {}, not RGB/RGBA'.format(base.mode))

    draw = ImageDraw.Draw(base)
    text = word_wrap(text, draw, base.width)
    text_width, text_height = draw.multiline_textsize(text, font=font)
    center_x = base.width // 2
    center_y = base.height // 2
    draw.multiline_text((center_x - text_width / 2, center_y - text_height / 2), text, font=font,
                        fill=white, stroke_width=1, stroke_fill=black, align='center')


def make_image(text: str, style_id: str):
    outfile = BytesIO()
    style = [s for s in STYLES if s.id == style_id][0]
    with Image.open(style.image) as base:
        render_text(text, base)
        base.save(outfile, 'PNG')
    return outfile.getvalue()


def make_video(text: str, style_id: str):
    ev = libhoney.Event()
    ev.add_field('type', 'bake')
    ev.add_field('text', text)
    ev.add_field('style', style_id)
    ev.send()
    style = [s for s in STYLES if s.id == style_id][0]
    with Image.open(style.image) as image:
        size = image.size
    overlay = Image.new('RGBA', size, (0, 0, 0, 0))
    render_text(text, overlay)
    with tempfile.TemporaryDirectory() as tmp:
        overlay_file = Path(tmp) / 'overlay.png'
        overlay.save(overlay_file)
        result_file = Path(tmp) / 'result.mp4'
        subprocess.run([
            'ffmpeg',
            '-hide_banner',
            '-v', 'warning',
            '-i', str(style.video),
            '-i', str(overlay_file),
            '-filter_complex', 'overlay=x=0:y=0',
            '-c:v', 'libx264', '-preset', 'superfast', '-crf', '27', '-f', 'mp4', '-c:a', 'copy',
            '-y', str(result_file)
        ]).check_returncode()
        return result_file.read_bytes()


def start(update: telegram.Update, context: CallbackContext):
    template = r"""Hi {you}!
To shitpost, type @{me} and type the text you want to overlay over crab rave.
This was originally made by @boringcactus in one afternoon when ze was bored.
This bot isn't super reliable but the source is at https://github.com/boringcactus/crabravebot,
and you can use this bot from the Web at https://{name}.herokuapp.com/"""
    you = update.effective_user.first_name
    if you is None or len(you) == 0:
        you = update.effective_user.username
    text = template.format(you=you, me=context.bot.username, name=os.environ.get('HEROKU_APP_NAME', 'crabravebot'))
    context.bot.send_message(chat_id=update.effective_chat.id, text=text)


def inline_query(update: telegram.Update, context: CallbackContext):
    query = update.inline_query
    text = query.query

    logger.info('Got query %s', text)

    def make_result(style: Style):
        base = "https://" + os.environ.get('HEROKU_APP_NAME', 'crabravebot') + ".herokuapp.com/render?v=1&style=" + style.id + "&text=" + url_quote(text, safe='')
        return telegram.InlineQueryResultVideo(
            id=style.id,
            video_url=base + '&ext=mp4',
            mime_type="video/mp4",
            thumb_url=base + '&ext=png',
            title=style.name,
        )

    query.answer([make_result(s) for s in STYLES])


def message(update: telegram.Update, context: CallbackContext):
    text = update.effective_message.text

    logger.info('Got query %s', text)

    update.effective_message.reply_video(BytesIO(make_video(text, STYLES[0].id)))


@app.route('/')
def index():
    return render_template('index.html', styles=STYLES)


@app.route('/render')
def serve_render():
    style = request.args['style']
    ext = request.args['ext']
    text = request.args['text']
    if sum(1 for x in STYLES if x.id == style) == 0:
        raise ValueError('bad style')
    if ext == 'png':
        response = make_image(text, style)
        content_type = 'image/png'
    elif ext == 'mp4':
        response = make_video(text, style)
        content_type = 'video/mp4'
    else:
        raise ValueError('bad extension')
    return response, {'Content-Type': content_type}


TOKEN = os.environ.get('TG_BOT_TOKEN')
WEBHOOK = '/webhook/' + TOKEN


@app.route(WEBHOOK, methods=['POST'])
def webhook():
    update_queue.put(telegram.Update.de_json(request.get_json(), bot))
    return 'ok'


bot = telegram.Bot(token=TOKEN)
bot.set_webhook('https://' + os.environ.get('HEROKU_APP_NAME', 'crabravebot') + '.herokuapp.com' + WEBHOOK)
update_queue = Queue()
dp = Dispatcher(bot, update_queue, use_context=True)
# Add handlers
dp.add_handler(CommandHandler('start', start))
dp.add_handler(InlineQueryHandler(inline_query))
dp.add_handler(MessageHandler(Filters.text, message))

thread = Thread(target=dp.start, name='dispatcher')
thread.start()
